### PR TITLE
[AIRFLOW-936] Add clear/mark success for DAG in the UI

### PR DESCRIPTION
**Pull request by allisonwang**
Dear Airflow maintainers,

Please accept this PR. I understand that it will not be reviewed until I have checked off all the steps below!

### JIRA
- [x] My PR addresses [Airflow 936](https://issues.apache.org/jira/browse/AIRFLOW-936) issues and references them in the PR title. 

### Description
- [x] Here are some details about my PR, including screenshots of any UI changes:

This PR adds a modal popup when clicking circle DAG icon in Airflow tree view UI. It adds the functionalities to clear/mark success of the entire DAG run. This behavior is equivalent to individually clear/mark each task instance in the DAG run. The original logic of editing DAG run page is moved to the button "Edit DAG Run".

screen shot 2017-06-07 at 10 19 59 am
screen shot 2017-06-01 at 10 42 14 am

### Tests
- [x] My PR adds the following unit tests __OR__ does not need testing for this extremely good reason:

 Tests are added under `test/api/common/mark_tasks.py` to test marking entire DAG as success. There is no test for clear the DAG since it is directly implemented in `views.py`.

### Commits
- [x] My commits all reference JIRA issues in their subject lines, and I have squashed multiple commits if they address the same issue. In addition, my commits follow the guidelines from "[How to write a good git commit message](http://chris.beams.io/posts/git-commit/)":
    1. Subject is separated from body by a blank line
    2. Subject is limited to 50 characters
    3. Subject does not end with a period
    4. Subject uses the imperative mood ("add", not "adding")
    5. Body wraps at 72 characters
    6. Body explains "what" and "why", not "how"

@aoen @saguziel

### CODE REVIEW

File: airflow/www/views.py [hunk 1]
```diff
@@ -1034,10 +1034,13 @@ def clear(self):
         past = request.args.get('past') == "true"
         recursive = request.args.get('recursive') == "true"
```
Comment (aoen) [inline airflow/www/views.py]: Nit: kill whitespace

File: airflow/api/common/experimental/mark_tasks.py [hunk 2]
```diff
@@ -181,7 +181,40 @@ def set_state(task, execution_date, upstream=False, downstream=False,
         if len(sub_dag_ids) > 0:
             tis_altered += qry_sub_dag.all()
 
+    session.expunge_all()
```
Comment (bolkedebruin) [inline airflow/api/common/experimental/mark_tasks.py:183]: This seems unrelated. Why add this?

File: airflow/api/common/experimental/mark_tasks.py [hunk 3]
```diff
@@ -181,7 +181,40 @@ def set_state(task, execution_date, upstream=False, downstream=False,
...
     session.close()
 
     return tis_altered
 
+def set_dag_state(dag, execution_date, state=State.SUCCESS, commit=False):
```
Comment (bolkedebruin) [inline airflow/api/common/experimental/mark_tasks.py]: You are not setting a dag state but a dagrun state please rename.

File: airflow/api/common/experimental/mark_tasks.py [hunk 4]
```diff
@@ -181,7 +181,40 @@ def set_state(task, execution_date, upstream=False, downstream=False,
...
+                              state=state, commit=commit)
+        res.extend(new_state)
+
+    # Mark the DAG run
+    if commit:
+        drs = DagRun.find(dag.dag_id, execution_date=execution_date)
+        if drs and len(drs) == 1:
+            dr = drs[0]
+            dr.dag = dag
+            dr.update_state()
```
Comment (amaliujia) [inline airflow/api/common/experimental/mark_tasks.py]: Add `else` and add logging say cannot find dag_run in DB or find more than one dag_run given dag_id and execution_date in DB?

File: airflow/api/common/experimental/mark_tasks.py (hunk 2 above)
Comment (allisonwang) [reply airflow/api/common/experimental/mark_tasks.py:183]: @bolkedebruin This is for appending results from this function. Without expunge sqlalchemy will through UnboundedExceptionError since the session is closed before function returns.

File: airflow/api/common/experimental/mark_tasks.py (hunk 4 above)
Comment (saguziel) [reply airflow/api/common/experimental/mark_tasks.py]: There can't be more than one dag_run
The schema for dag_run has
```PRIMARY KEY (`id`),
  UNIQUE KEY `dag_id` (`dag_id`,`execution_date`),
  UNIQUE KEY `dag_id_2` (`dag_id`,`run_id`),```

So it should be either 0 or 1

**Comment by saguziel** [comment]
I haven't really dived in, but does it make sense that clear and success are separate endpoints?

File: airflow/api/common/experimental/mark_tasks.py (hunk 4 above)
Comment (amaliujia) [reply airflow/api/common/experimental/mark_tasks.py]: I think more checks are good.

File: tests/api/common/mark_tasks.py [hunk 5]
```diff
@@ -207,5 +208,189 @@ def tearDown(self):
...
+    def setUp(self):
+        self.dagbag = models.DagBag(include_examples=True)
+        self.dag1 = self.dagbag.dags['test_example_bash_operator']
+        self.dag2 = self.dagbag.dags['example_subdag_operator']
+
+        self.execution_dates = [days_ago(3), days_ago(2), days_ago(1)]
+
+        self.session = Session()
+
+    def verify_dag_states(self, dag, date, state=State.SUCCESS):
```
Comment (amaliujia) [inline tests/api/common/mark_tasks.py]: Same as above. You are testing dagrun. So can rename tests.

File: airflow/api/common/experimental/mark_tasks.py (hunk 4 above)
Comment (allisonwang) [reply airflow/api/common/experimental/mark_tasks.py]: Good to know! I will make a note in `else` branch.

Comment (saguziel) [reply airflow/api/common/experimental/mark_tasks.py]: I feel like it's just code bloat this point given it's impossible to get multiple results. I think the better thing to do is just loop over the results

File: airflow/api/common/experimental/mark_tasks.py [hunk 6]
```diff
@@ -181,7 +181,39 @@ def set_state(task, execution_date, upstream=False, downstream=False,
...
     return tis_altered
 
+def set_dag_run_state(dag, execution_date, state=State.SUCCESS, commit=False):
+    """
+    Set the state of a dag run and all task instances associated with the dag
+    run for a specific execution date.
+    :param dag: the DAG of which to alter state
+    :param execution_date: the execution date from which to start looking
+    :param state: the state to which the DAG need to be set
+    :param commit: commit DAG and tasks to be altered to the database
```
Comment (aoen) [inline airflow/api/common/experimental/mark_tasks.py:195]: The name of this method seems to imply that commit should always be true. Maybe there should be another method that gets a list of task instances that would be modified without altering them?

File: airflow/www/views.py [hunk 7]
```diff
@@ -1034,10 +1034,13 @@ def clear(self):
...
 
-        dag = dag.sub_dag(
-            task_regex=r"^{0}$".format(task_id),
-            include_downstream=downstream,
-            include_upstream=upstream)
+        clear_all = request.args.get('clear_all') == "true"
+
+        if not clear_all:
```
Comment (aoen) [inline airflow/www/views.py]: This should be a different endpoint for dagruns/tasks, although they could reuse most of the same logic.

File: airflow/www/templates/airflow/dag.html [hunk 8]
```diff
@@ -216,6 +216,36 @@ <h4 class="modal-title" id="myModalLabel">
       </div>
     </div>
   </div>
+  <!-- Modal for dag -->
```
Comment (aoen) [inline airflow/www/templates/airflow/dag.html:219]: "Edit"
"Clear"
and
"Mark Success"

are fine for the button names.

File: airflow/api/common/experimental/mark_tasks.py [hunk 9]
```diff
@@ -13,16 +13,16 @@
 # limitations under the License.
 
 import datetime
+import logging
```
Comment (saguziel) [inline airflow/api/common/experimental/mark_tasks.py]: This import appears unused

File: airflow/api/common/experimental/mark_tasks.py [hunk 10]
```diff
@@ -13,16 +13,16 @@
...
 
 from airflow.jobs import BackfillJob
-from airflow.models import DagRun, TaskInstance
+from airflow.models import DagRun, DagStat, TaskInstance
```
Comment (saguziel) [inline airflow/api/common/experimental/mark_tasks.py]: This import also appears unused

**Comment by saguziel** [comment]
It seems that if clear_dagrun can be part of clear, dag_success could be part of success

**Comment by saguziel** [comment]
And maybe it should be dagrun_success

File: airflow/www/templates/airflow/dag.html [hunk 11]
```diff
@@ -342,6 +396,15 @@ <h4 class="modal-title" id="myModalLabel">
       window.location = url;
     });
 
+    $('#btn_dag_success').click(function(){
```
Comment (aoen) [inline airflow/www/templates/airflow/dag.html]: Nit: replace "dag" with "dagrun"

File: airflow/www/templates/airflow/dag.html [hunk 12]
```diff
@@ -367,5 +430,9 @@ <h4 class="modal-title" id="myModalLabel">
       $.post(url);
     });
 
+    $('#btn_edit_dag').click(function(){
```
Comment (aoen) [inline airflow/www/templates/airflow/dag.html]: Nit: edit_dagrun

**Comment by aoen** [comment]
@saguziel I think dagrun/task clearing should be separate functions, having them in the same function is the equivalent to having 15 different classes and then having a function that switches off of the class type, instead of just having 15 different functions to call in the first place.

**Comment by saguziel** [comment]
@aoen That is a pattern though. We save some code duplication, and we don't have 15 classes. Especially since we aren't actually clearing the dag_run, but rather the tasks that belong to it.

File: airflow/www/views.py [hunk 13]
```diff
@@ -1093,6 +1096,46 @@ def blocked(self):
...
+            return redirect(origin)
+
+        execution_date = dateutil.parser.parse(execution_date)
+        dag = dagbag.get_dag(dag_id)
+
+        if not dag:
+            flash('Cannot find DAG: {}'.format(dag_id), 'error')
+            return redirect(origin)
+
+        from airflow.api.common.experimental.mark_tasks import set_dag_run_state
```
Comment (saguziel) [inline airflow/www/views.py]: Do a top level import instead

File: airflow/api/common/experimental/mark_tasks.py [hunk 14]
```diff
@@ -181,7 +181,39 @@ def set_state(task, execution_date, upstream=False, downstream=False,
         if len(sub_dag_ids) > 0:
             tis_altered += qry_sub_dag.all()
 
+    session.expunge_all()
     session.close()
 
     return tis_altered
 
+def set_dag_run_state(dag, execution_date, state=State.SUCCESS, commit=False):
```
Comment (saguziel) [inline airflow/api/common/experimental/mark_tasks.py:188]: Actually, put this in models.py. It's okay to have the stub here but the logic of setting the dagrun to the state should be in models

**Comment by aoen** [comment]
Spoke with @saguziel offline, we are in agreement that dagrun/task clearing should be separate functions that call some common functions.

File: airflow/www/views.py (hunk 7 above)
Comment (saguziel) [reply airflow/www/views.py]: Should be able to call clear_task from clear_dagrun,  likewise for success

File: airflow/api/common/experimental/mark_tasks.py (hunk 14 above)
Comment (allisonwang) [reply airflow/api/common/experimental/mark_tasks.py:188]: @saguziel There is a function called `set_dag_run_state` in `models.py`, but the logics are quite different. `set_dag_run_state` here basically called `set_state` above which is designed for mark success functionality and considers a lot more edge cases. But definitely the API functions need refactoring in the future to be consistent with other parts of the codebase.

File: airflow/www/templates/airflow/tree.html [hunk 15]
```diff
@@ -1,13 +1,13 @@
-{# 
+{#
   Licensed to the Apache Software Foundation (ASF) under one or more
   contributor license agreements.  See the NOTICE file distributed with
   this work for additional information regarding copyright ownership.
   The ASF licenses this file to You under the Apache License, Version 2.0
   (the "License"); you may not use this file except in compliance with
   the License.  You may obtain a copy of the License at
-  
+
```
Comment (aoen) [inline airflow/www/templates/airflow/tree.html:8]: Looks like these were accidental removals

File: airflow/www/views.py [hunk 16]
```diff
@@ -1070,6 +1071,47 @@ def clear(self):
...
+        origin = request.args.get('origin')
+        execution_date = request.args.get('execution_date')
+        confirmed = request.args.get('confirmed') == "true"
+
+        dag = dagbag.get_dag(dag_id)
+        execution_date = dateutil.parser.parse(execution_date)
+        start_date = execution_date
+        end_date = execution_date
+
+        if confirmed:
```
Comment (aoen) [inline airflow/www/views.py]: This section can be factored out with the /clear endpoint code.

also I think the recursive flag needs to be supported here too for the subdag operator (same as /clear).

Comment (allisonwang) [reply airflow/www/views.py]: @aoen  clear here default to include subdags https://github.com/apache/incubator-airflow/blob/master/airflow/models.py#L3149 but adding it definitely makes it more readable :) Will refactor out the logic.

Comment (aoen) [reply airflow/www/views.py]: As far as I understand we should allow the option to specify recursive the same way it can be specified for task-clearing (for the subdag operator), i.e. I think it should be an option in the modal.

File: airflow/www/templates/airflow/tree.html (hunk 15 above)
Comment (allisonwang) [reply airflow/www/templates/airflow/tree.html:8]: @aoen Hmm maybe my editor (Atom) removed the white space automatically but the Apache license content should be exactly the same.

File: airflow/www/views.py [hunk 17]
```diff
@@ -1041,34 +1071,27 @@ def clear(self):
...
+        origin = request.args.get('origin')
+        execution_date = request.args.get('execution_date')
+        confirmed = request.args.get('confirmed') == "true"
+
+        dag = dagbag.get_dag(dag_id)
+        execution_date = dateutil.parser.parse(execution_date)
+        start_date = execution_date
+        end_date = execution_date
+
+        return self._clear_dag_tis(dag, start_date, end_date, True, origin, confirmed)
```
Comment (aoen) [inline airflow/www/views.py]: it's better to use keyword arguments like recursive=True here (that way it's clear what each argument is). The recursive value should not always be True but have it's value derived from the recursive button in the model, same as for individual task instance clearing. LGTM after that though!

Comment (allisonwang) [reply airflow/www/views.py]: @aoen Right we can make the clear DAG run UI more consistent with clear tasks, i.e. support past / future, recursive/ non-recursive, etc. Here I feel letting recursive always be true make sense otherwise clear DAG run will clear all operator task instances except for subdag which could be confusing.

Comment (allisonwang) [reply airflow/www/views.py]: Maybe in version 2 :)

Comment (aoen) [reply airflow/www/views.py]: I think past/future are a bit different than non-recursive (which set of dagruns are affected vs how affected tasks are changed), but I am OK leaving the recursive option as a TODO (it's just not that great that clearing dagruns doesn't have functional parity with clearing task instances).

**Comment by aoen** [comment]
LGTM
